### DISCUSSION TITLE
tsrpc工程中没有指定传输协议POST/GET，源码中也没有传输协议，这是怎么实现的

### DISCUSSION THREAD

**Discussion by Jonnypeng** [k8w/tsrpc#8] [Q&A] [answered]
在整个工程中，获取数据与更新数据都没有涉及到POST/GET的声明，请问是不是RPC都不用去声明，不明白工程中的浏览器环境下的GET/POST传输在工程中是如何实现的

**Comment by k8w** [comment, marked as answer]
因为 TSRPC 在一开始就设计为 **传输协议无关的**。

所以你既可以在 HTTP 协议下使用，也可以在 WebSocket 下使用，也可以方便的扩展到原生 TCP、UDP 等协议。
试想一下，在 WebSocket 下自然是没有 POST/GET 的吧，所以对于传输协议无关的上层配置，自然不需要关注配置这些细节。

具体实现层面，HTTP Server 是统一使用 POST 的，当然如有需要，你可以通过 [Flow](https://tsrpc.cn/docs/flow/custom-http-response) 去扩展实现 GET 协议。

**Comment by hooshuo** [comment]
扩展原有的框架，现在要加一个支持Http Post的服务，有没有什么办法实现？如果是通过Flow扩展实现，post传的数据从哪里获得。

**Reply by k8w** [reply]
preRecvDataFlow 中
conn.httpReq conn.httpRes 是 NodeJS 原始的 req res

参见例子 自定义 HTTP 响应

**Reply by k8w** [reply]
v.data 就是原始数据
